File: java-algo/src/main/java/com/leetcode/Problem1038.java
package com.leetcode;

import com.common.TreeNode;

/**
 * 1038. Binary Search Tree to Greater Sum Tree
 * https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/
 */
public class Problem1038 {
    private int sum = 0;

    public TreeNode bstToGst(TreeNode root) {
        if (root == null) {
            return null;
        }
        bstToGst(root.right);
        sum += root.val;
        root.val = sum;
        bstToGst(root.left);
        return root;
    }
}

File: python-algo/leetcode/problem_1038.py
"""1038. Binary Search Tree to Greater Sum Tree
https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/
"""
from common.tree_node import TreeNode


class Solution:
    def bst_to_gst(self, root: TreeNode) -> TreeNode:
        p = root
        stack, node_seq = [], []
        # reversed in-order traversal.
        while p or stack:
            if p:
                stack.append(p)
                p = p.right
            else:
                node = stack.pop()
                node_seq.append(node)
                p = node.left
        pre_sum = 0
        for node in node_seq:
            pre_sum += node.val
            node.val = pre_sum
        return root
